File: test/base/JsonDeserializer.java
package base;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonElement;
import play.Logger;

import java.lang.reflect.Type;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

final class JsonDeserializer {

    private JsonDeserializer() {
    }

    private static final GsonBuilder gsonBuilder = new GsonBuilder();
    static {
        gsonBuilder.registerTypeAdapter(Date.class, new DateDeserializer());
    }
    private static final Gson gson = gsonBuilder.create();

    private static class DateDeserializer implements com.google.gson.JsonDeserializer<Date> {

        @Override
        public Date deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
            try {
                return new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").parse(json.getAsString());
            } catch (ParseException e) {
                try {
                    return new Date(json.getAsLong());
                } catch (RuntimeException e2) {
                    Logger.warn("Failed to parse date " + json.getAsString());
                }
            }
            return null;
        }
    }

    static <T> T deserialize(Class<T> model, JsonNode node) {
        return gson.fromJson(node.toString(), model);
    }
}

File: app/util/json/JsonDeserializer.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package util.json

import com.fasterxml.jackson.databind.JsonNode
import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import org.joda.time.DateTime
import org.joda.time.format.ISODateTimeFormat
import play.api.Logging

import java.lang.reflect.Type
import java.text.SimpleDateFormat
import java.util.Date
import scala.util.{Success, Try}

object JsonDeserializer:
  private val builder = new GsonBuilder()
  builder.registerTypeAdapter(classOf[Date], new DateDeserializer())
  builder.registerTypeAdapter(classOf[DateTime], new DateTimeDeserializer())
  private val gson = builder.create()

  def deserialize[T](model: Class[T], node: JsonNode): T = gson.fromJson(node.toString, model)

class DateDeserializer extends JsonDeserializer[Date] with Logging:
  override def deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Date =
    Try(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").parse(json.getAsString)) match
      case Success(d) => d
      case _ =>
        Try(new Date(json.getAsLong)) match
          case Success(dt) => dt
          case _ =>
            Try(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX").parse(json.getAsString)) match
              case Success(dt) => dt
              case _ =>
                logger.warn(s"Failed to parse date ${json.getAsString}")
                null

class DateTimeDeserializer extends JsonDeserializer[DateTime] with Logging:
  override def deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): DateTime =
    Try(ISODateTimeFormat.dateTime().parseDateTime(json.getAsString)) match
      case Success(dt) => dt
      case _ =>
        Try(new DateTime(json.getAsLong)) match
          case Success(dt) => dt
          case _ =>
            logger.warn(s"Failed to parse datetime ${json.getAsString}")
            null
